core/src/main/java/org/infinispan/util/concurrent/CommandAckCollector.java

```diff
@@ -0,0 +1,512 @@
+      SingleKeyCollector collector = (SingleKeyCollector) collectorMap.get(id);
+      if (collector != null) {
+         collector.backupAck(topologyId, from);
+      }
+   }
+
+   /**
+    * Acknowledges a write operation completion in the primary owner.
+    * <p>
+    * If the operation does not succeed (conditional commands), the collector is completed without waiting for the
+    * acknowledges from the backup owners.
+    *
+    * @param id          the {@link CommandInvocationId}.
+    * @param returnValue the return value.
+    * @param success     {@code true} if the operation succeed in the primary owner, {@code false} otherwise.
+    * @param from        the primary owner.
+    * @param topologyId  the topology id.
+    */
+   public void primaryAck(CommandInvocationId id, Object returnValue, boolean success, Address from, int topologyId) {
+      SingleKeyCollector collector = (SingleKeyCollector) collectorMap.get(id);
+      if (collector != null) {
+         collector.primaryAck(topologyId, returnValue, success, from);
+      }
+   }
+
+   /**
+    * Acknowledges an exception during the operation execution.
+    * <p>
+    * The collector is completed without waiting any further acknowledges.
+    *
+    * @param id         the {@link CommandInvocationId}.
+    * @param throwable  the {@link Throwable}.
+    * @param topologyId the topology id.
+    */
+   public void completeExceptionally(CommandInvocationId id, Throwable throwable, int topologyId) {
+      Collector<?> collector = collectorMap.get(id);
+      if (collector != null) {
+         collector.completeExceptionally(throwable, topologyId);
+      }
+   }
+
+   /**
+    * Waits for all the acknowledges from the primary and backup owners affected by the operation.
+    * <p>
+    * The collector is removed after completion or timeout.
+    *
+    * @param id          the {@link CommandInvocationId}.
+    * @param timeout     the maximum time to wait.
+    * @param timeUnit    the time unit of the timeout argument.
+    * @param returnValue the local operation's return value.
+    * @param <T>         the return value type.
+    * @return the operation return value from the primary owner or {@code returnValue} if the collector does not exists.
+    * @throws Throwable an exception if an error occurred or a timeout happens.
+    */
+   public <T> T awaitCollector(CommandInvocationId id, long timeout, TimeUnit timeUnit, T returnValue) throws Throwable {
+      //noinspection unchecked
+      Collector<T> collector = (Collector<T>) collectorMap.get(id);
+      if (collector != null) {
+         try {
+            if (trace) {
+               log.tracef("[Collector#%s] Waiting for acks.", id);
+            }
+            return awaitFuture(collector.getFuture(), timeout, timeUnit);
+         } catch (CompletionException e) {
+            throw CompletableFutures.extractException(e);
+         } finally {
+            collectorMap.remove(id);
```

Comment:
I don't see the advantage of it. I need to create a collector every time the command is retried. So, I need to remove it when the command finishes or when the command needs to retry.